File: packages/api/src/models/SharedExport.ts
/**
 * SharedExport Model — Public shareable links for annotation exports
 *
 * Stores pre-rendered export content with a unique token so that
 * anyone with the link can view the export without authentication.
 */

import { DataTypes, Model, Optional } from 'sequelize';
import { sequelize } from '../db';

interface SharedExportAttributes {
  id: string;
  userId: string;
  bookId: string;
  token: string;
  format: string;
  title: string;
  content: string;
  contentType: string;
  viewCount: number;
  expiresAt: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

interface SharedExportCreationAttributes extends Optional<SharedExportAttributes, 'id' | 'viewCount' | 'expiresAt' | 'createdAt' | 'updatedAt'> {}

export class SharedExport extends Model<SharedExportAttributes, SharedExportCreationAttributes> implements SharedExportAttributes {
  public id!: string;
  public userId!: string;
  public bookId!: string;
  public token!: string;
  public format!: string;
  public title!: string;
  public content!: string;
  public contentType!: string;
  public viewCount!: number;
  public expiresAt!: Date | null;
  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
}

SharedExport.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'users',
        key: 'id',
      },
      onDelete: 'CASCADE',
    },
    bookId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'books',
        key: 'id',
      },
      onDelete: 'CASCADE',
    },
    token: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true,
    },
    format: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    title: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    content: {
      type: DataTypes.TEXT,
      allowNull: false,
    },
    contentType: {
      type: DataTypes.STRING,
      allowNull: false,
      defaultValue: 'text/markdown; charset=utf-8',
    },
    viewCount: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    expiresAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    createdAt: {
      type: DataTypes.DATE,
      allowNull: false,
    },
    updatedAt: {
      type: DataTypes.DATE,
      allowNull: false,
    },
  },
  {
    sequelize,
    tableName: 'shared_exports',
    timestamps: true,
    indexes: [
      { fields: ['token'], unique: true },
      { fields: ['user_id'] },
      { fields: ['book_id'] },
    ],
  }
);

File: packages/server/app/models/shared_export.py
"""Shared export model."""

import uuid
from datetime import datetime, UTC
from typing import TYPE_CHECKING
from uuid import UUID

from sqlalchemy import DateTime, ForeignKey, Index, Integer, String, Text, text
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.db import Base

if TYPE_CHECKING:
    from app.models.book import Book
    from app.models.user import User


class SharedExport(Base):
    __tablename__ = 'shared_exports'
    __table_args__ = (
        Index('ix_shared_exports_user_book', 'user_id', 'book_id'),
        Index('ix_shared_exports_expires_at', 'expires_at'),
    )

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        server_default=text('gen_random_uuid()'),
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('users.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    book_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('books.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    token: Mapped[str] = mapped_column(
        String(255),
        unique=True,
        nullable=False,
        index=True,
    )
    format: Mapped[str] = mapped_column(String(255), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    content_type: Mapped[str] = mapped_column(
        String(255),
        default='text/markdown; charset=utf-8',
    )
    view_count: Mapped[int] = mapped_column(Integer, default=0)
    expires_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=lambda ctx: datetime.now(tz=UTC),
    )

    user: Mapped['User'] = relationship('User', back_populates='shared_exports')
    book: Mapped['Book'] = relationship('Book', back_populates='shared_exports')
